iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp

```diff
@@ -0,0 +1,333 @@
+      if (!val) return failure();
+      vecArgs.push_back(val);
+    }
+    auto vecType = getVecType(rewriter, op.getResult().getType());
+    if (!vecType) return failure();
+    auto newOp = rewriter.create<OpTy>(op.getLoc(), vecType, vecArgs);
+    rewriter.replaceOp(op, newOp.getOperation()->getResults());
+    return success();
+  }
+};
+
+template <typename OpTy>
+struct VectorizeCmpOp : public OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      OpTy op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (llvm::any_of(args,
+                     [](Value arg) { return arg.getType().isIntOrFloat(); })) {
+      return failure();
+    }
+    auto newOp =
+        rewriter.create<OpTy>(op.getLoc(), op.predicate(), args[0], args[1]);
+    rewriter.replaceOp(op, newOp.getResult());
+    return success();
+  }
+};
+
+struct VectorizeSelectOp : public OpConversionPattern<mlir::SelectOp> {
+  using OpConversionPattern<mlir::SelectOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      mlir::SelectOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (llvm::any_of(args,
+                     [](Value arg) { return arg.getType().isIntOrFloat(); })) {
+      return failure();
+    }
+    auto newOp =
+        rewriter.create<SelectOp>(op.getLoc(), args[0], args[1], args[2]);
+    rewriter.replaceOp(op, newOp.getResult());
+    return success();
+  }
+};
+
+struct VectorizeYieldOp : public OpConversionPattern<linalg::YieldOp> {
+  using OpConversionPattern<linalg::YieldOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      linalg::YieldOp yieldOp, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (!llvm::all_of(args,
+                      [](Value v) { return v.getType().isa<VectorType>(); }))
+      return failure();
+    SmallVector<Value, 4> vecArgs;
+    for (Value arg : args) {
+      Value val = legalizeToVectorType(rewriter, arg);
+      if (!val) return failure();
+      vecArgs.push_back(val);
+    }
+    rewriter.replaceOpWithNewOp<linalg::YieldOp>(yieldOp, vecArgs);
+    return success();
+  }
+};
+
+struct VectorizeGenericOp : public OpConversionPattern<linalg::GenericOp> {
+  using OpConversionPattern<linalg::GenericOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      linalg::GenericOp genericOp, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    // Ignore a case that either a nop pattern or a linalg.fill pattern.
+    if (isa<linalg::YieldOp>(
+            genericOp.region().front().getOperations().front())) {
+      return failure();
+    }
+
+    // Do not vectorize if one of the operand is 0-D or one of the operand is
+    // not iterated on contiguous memory.
+    for (auto map : genericOp.getIndexingMaps()) {
```

Comment:
We also need to check inner most iterator is parallel

Reply:
It expects to operate on element-wise pass, so check if all the loops are parallel.